File: VrchAndroid/app/src/main/java/nokamoto/github/com/vrchandroid/firebase/FirebaseMessageClient.java
package nokamoto.github.com.vrchandroid.firebase;

import android.util.Log;

import com.google.android.gms.tasks.Tasks;
import com.google.common.base.Optional;
import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;

public class FirebaseMessageClient {
    private final static String TAG = FirebaseMessageClient.class.getSimpleName();
    private final static String MESSAGES = "messages";
    private final static String ROOMS = "rooms";

    private FirebaseDatabase database;
    private ChildEventListener listener;

    private String room;
    private String p;

    public FirebaseMessageClient(String room) {
        this.room = room;
        this.database = FirebaseDatabase.getInstance();
        p = "#" + room + " ";
    }

    private DatabaseReference roomRef() {
        return database.getReference().child(ROOMS).child(room);
    }

    private DatabaseReference messageRef() {
        return database.getReference().child(MESSAGES).child(room);
    }

    public interface MessageEventListener {
        void onAdded(FirebaseMessage message);
    }

    public interface RoomValueListener {
        void onValue(FirebaseRoom room);
        void onCancelled();
    }

    public void onStart(final MessageEventListener listener, final RoomValueListener valueListener) {
        this.listener = new ChildEventListener() {
            @Override
            public void onChildAdded(DataSnapshot dataSnapshot, String s) {
                Optional<FirebaseMessage> data = FirebaseMessage.fromSnapshot(dataSnapshot);
                if (data.isPresent()) {
                    listener.onAdded(data.get());
                } else {
                    Log.e(TAG, p + "unexpected data snapshot: " + dataSnapshot);
                }
            }

            @Override
            public void onChildChanged(DataSnapshot dataSnapshot, String s) {

            }

            @Override
            public void onChildRemoved(DataSnapshot dataSnapshot) {

            }

            @Override
            public void onChildMoved(DataSnapshot dataSnapshot, String s) {

            }

            @Override
            public void onCancelled(DatabaseError databaseError) {
                Log.e(TAG, p + "cancelled.", databaseError.toException());
            }
        };

        messageRef().
                orderByChild(FirebaseMessage.CREATED_AT).
                startAt(System.currentTimeMillis() - TimeUnit.DAYS.toMillis(7)).
                addChildEventListener(this.listener);

        roomRef().addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {
                valueListener.onValue(FirebaseRoom.fromSnapshot(dataSnapshot).orNull());
            }

            @Override
            public void onCancelled(DatabaseError databaseError) {
                Log.i(TAG, p + "cancelled.", databaseError.toException());
                valueListener.onCancelled();
            }
        });
    }

    public void onStop() {
        if (listener != null) {
            messageRef().removeEventListener(listener);
        }
    }

    public void write(FirebaseMessage message) throws ExecutionException, InterruptedException {
        Log.i(TAG, p + "write message: " + message);

        String pushed = messageRef().push().getKey();
        Map<String, Object> updates = new HashMap<>();
        updates.put(pushed, message.toMap());

        Tasks.await(messageRef().updateChildren(updates));
    }

    public void write(FirebaseRoom room) throws ExecutionException, InterruptedException {
        Log.i(TAG, p + "write room: " + room);
        Tasks.await(roomRef().setValue(room.toMap()));
    }
}

File: slackbridge/src/main/scala/vrch/slackbridge/firebase/FirebaseMessageClient.scala
package vrch.slackbridge.firebase

import java.io.FileInputStream

import com.google.firebase.auth.FirebaseCredentials
import com.google.firebase.database._
import com.google.firebase.tasks.{OnCompleteListener, Task}
import com.google.firebase.{FirebaseApp, FirebaseOptions}

import scala.concurrent.{Future, Promise}

class FirebaseMessageClient(config: FirebaseConfig, room: String) extends vrch.Logger {
  import config._

  private[this] val serviceAccount = new FileInputStream(json)

  private[this] val options = new FirebaseOptions.Builder()
    .setCredential(FirebaseCredentials.fromCertificate(serviceAccount))
    .setDatabaseUrl(url)
    .build()

  FirebaseApp.initializeApp(options)

  private def roomRef(): DatabaseReference = FirebaseDatabase.getInstance().getReference(s"/rooms/$room")

  private def messageRef(): DatabaseReference = FirebaseDatabase.getInstance().getReference(s"/messages/$room")

  def context(): Future[FirebaseRoom] = {
    val promise = Promise[FirebaseRoom]

    roomRef().addListenerForSingleValueEvent(new ValueEventListener {
      override def onCancelled(error: DatabaseError): Unit = promise.failure(error.toException)

      override def onDataChange(snapshot: DataSnapshot): Unit = {
        try {
          promise.success(FirebaseRoom(snapshot))
        } catch {
          case e: Exception => promise.failure(e)
        }
      }
    })

    promise.future
  }

  def addListener(callback: FirebaseMessage => Unit): Unit = {
    messageRef().orderByChild(FirebaseMessage.CREATED_AT).startAt(System.currentTimeMillis()).addChildEventListener(
      new ChildEventListener {
        override def onCancelled(error: DatabaseError): Unit = logger.error("cancelled.", error.toException)

        override def onChildChanged(snapshot: DataSnapshot, previousChildName: String): Unit = ()

        override def onChildMoved(snapshot: DataSnapshot, previousChildName: String): Unit = ()

        override def onChildAdded(snapshot: DataSnapshot, previousChildName: String): Unit = {
          try {
            val message = FirebaseMessage(snapshot)

            logger.info(s"added: $message")

            callback(message)
          } catch {
            case e: Exception => logger.error("on child added failed.", e)
          }
        }

        override def onChildRemoved(snapshot: DataSnapshot): Unit = ()
      }
    )
  }

  def write(message: FirebaseMessage): Future[Unit] = {
    logger.info(s"write: $message")

    val promise = Promise[Unit]
    val pushed = messageRef().push().getKey

    messageRef().child(pushed).setValue(message.toMap).addOnCompleteListener(new OnCompleteListener[Void] {
      override def onComplete(task: Task[Void]): Unit = {
        if (task.isSuccessful) promise.success(()) else promise.failure(task.getException)
      }
    })

    promise.future
  }
}
